File: modules/FaceDetection/ios/FaceDetectionFrameProcessor.swift
import CoreMedia
import NitroModules
import Vision
import VisionCamera

class FaceDetectionFrameProcessor: HybridFaceDetectionFrameProcessorSpec {
  private lazy var request: VNDetectFaceLandmarksRequest = {
    let r = VNDetectFaceLandmarksRequest()
    r.revision = VNDetectFaceLandmarksRequestRevision3
    return r
  }()
  private lazy var handler = VNSequenceRequestHandler()

  private let pipeline = FaceCapturePipeline()

  private var pendingCapture: [String: Any]? = nil
  private let pendingLock = NSLock()
  private var lastTargetPose: String? = nil

  override init() {
    super.init()
    pipeline.onCapture = { [weak self] result in
      self?.pendingLock.lock()
      self?.pendingCapture = result
      self?.pendingLock.unlock()
    }
  }

  func processFrame(frame: any HybridFrameSpec, args: FaceDetectionArgs) throws -> Variant_NullType_GuidanceResult_CaptureResult {
    let targetPose = args.targetPose
    if let previousPose = lastTargetPose, previousPose != targetPose {
      // Pose step changed (e.g. center -> left). Drop stale pending captures and
      // restart the native stabilization window for the new target.
      pipeline.reset()
      pendingLock.lock()
      pendingCapture = nil
      pendingLock.unlock()
    }
    lastTargetPose = targetPose

    pendingLock.lock()
    if let pending = pendingCapture {
      pendingCapture = nil
      pendingLock.unlock()
      // Drop stale captures from a previous pose step.
      let pendingPose = pending["poseId"] as? String
      if pendingPose == targetPose {
        return .third(captureResult(from: pending))
      }
    } else {
      pendingLock.unlock()
    }

    guard let nativeFrame = frame as? NativeFrame,
          let sampleBuffer = nativeFrame.sampleBuffer
    else {
      return noFaceResult()
    }

    do {
      try handler.perform([request], on: sampleBuffer, orientation: .leftMirrored)
    } catch {
      return noFaceResult(sampleBuffer: sampleBuffer, targetPose: targetPose)
    }

    guard let obs = request.results?.first as? VNFaceObservation else {
      return noFaceResult(sampleBuffer: sampleBuffer, targetPose: targetPose)
    }

    let bb = obs.boundingBox
    let bbX = Double(bb.minX)
    let bbY = Double(1.0 - bb.maxY)
    let bbW = Double(bb.width)
    let bbH = Double(bb.height)
    let cx = Float(bbX + bbW / 2.0)
    let cy = Float(bbY + bbH / 2.0)
    let faceSizeRatio = Float(bbH)

    let yaw = Float(obs.yaw.map { -$0.doubleValue * 180.0 / .pi } ?? 0.0)
    let pitch = Float(obs.pitch.map { $0.doubleValue * 180.0 / .pi } ?? 0.0)
    let roll = obs.roll.map { $0.doubleValue * 180.0 / .pi } ?? 0.0

    let faceRect = CGRect(x: bbX, y: bbY, width: bbW, height: bbH)
    let (rawBrightness, rawSharpness) = qualityMetrics(buffer: sampleBuffer, faceRect: faceRect)

    let progress = pipeline.processFrame(
      sampleBuffer: sampleBuffer,
      faceDetected: true,
      yaw: yaw,
      pitch: pitch,
      cx: cx,
      cy: cy,
      faceSizeRatio: faceSizeRatio,
      rawBrightness: Float(rawBrightness),
      rawSharpness: Float(rawSharpness),
      targetPose: targetPose
    )

    let guidance = GuidanceResult(
      type: .guidance,
      faceDetected: true,
      boundingBoxX: bbX,
      boundingBoxY: bbY,
      boundingBoxWidth: bbW,
      boundingBoxHeight: bbH,
      yaw: Double(yaw),
      pitch: Double(pitch),
      roll: roll,
      brightness: rawBrightness,
      sharpness: rawSharpness,
      faceSizeRatio: Double(faceSizeRatio),
      faceCenterX: Double(cx),
      faceCenterY: Double(cy),
      stabilizationProgress: Double(progress)
    )
    return .second(guidance)
  }

  private func captureResult(from payload: [String: Any]) -> CaptureResult {
    let poseId = payload["poseId"] as? String ?? "center"
    let uri = payload["uri"] as? String
    let qualityScore = payload["qualityScore"] as? Double ?? 0.0

    let rawScores = payload["scores"] as? [String: Any] ?? [:]
    let scores = CaptureScores(
      brightness: rawScores["brightness"] as? Double ?? 0.0,
      sharpness: rawScores["sharpness"] as? Double ?? 0.0,
      centeredness: rawScores["centeredness"] as? Double ?? 0.0,
      poseAccuracy: rawScores["poseAccuracy"] as? Double ?? 0.0,
      stability: rawScores["stability"] as? Double ?? 0.0,
      faceSize: rawScores["faceSize"] as? Double ?? 0.0,
      composite: rawScores["composite"] as? Double ?? qualityScore
    )

    return CaptureResult(
      type: .captured,
      poseId: poseId,
      uri: uri,
      qualityScore: qualityScore,
      scores: scores
    )
  }

  private func noFaceResult(sampleBuffer: CMSampleBuffer? = nil, targetPose: String = "center") -> Variant_NullType_GuidanceResult_CaptureResult {
    if let sampleBuffer {
      _ = pipeline.processFrame(
        sampleBuffer: sampleBuffer,
        faceDetected: false,
        yaw: 0,
        pitch: 0,
        cx: 0.5,
        cy: 0.5,
        faceSizeRatio: 0,
        rawBrightness: 0.5,
        rawSharpness: 0,
        targetPose: targetPose
      )
    }

    let guidance = GuidanceResult(
      type: .guidance,
      faceDetected: false,
      boundingBoxX: nil,
      boundingBoxY: nil,
      boundingBoxWidth: nil,
      boundingBoxHeight: nil,
      yaw: nil,
      pitch: nil,
      roll: nil,
      brightness: nil,
      sharpness: nil,
      faceSizeRatio: nil,
      faceCenterX: nil,
      faceCenterY: nil,
      stabilizationProgress: 0.0
    )
    return .second(guidance)
  }

  private func qualityMetrics(buffer: CMSampleBuffer, faceRect: CGRect) -> (brightness: Double, sharpness: Double) {
    guard let pixelBuffer = CMSampleBufferGetImageBuffer(buffer) else {
      return (0.5, 0.0)
    }
    CVPixelBufferLockBaseAddress(pixelBuffer, .readOnly)
    defer { CVPixelBufferUnlockBaseAddress(pixelBuffer, .readOnly) }

    guard CVPixelBufferGetPlaneCount(pixelBuffer) >= 1,
          let yBase = CVPixelBufferGetBaseAddressOfPlane(pixelBuffer, 0)
    else {
      return (0.5, 0.0)
    }

    let yPtr = yBase.assumingMemoryBound(to: UInt8.self)
    let bpr = CVPixelBufferGetBytesPerRowOfPlane(pixelBuffer, 0)
    let planeW = CVPixelBufferGetWidthOfPlane(pixelBuffer, 0)
    let planeH = CVPixelBufferGetHeightOfPlane(pixelBuffer, 0)

    let x0 = max(1, Int(faceRect.minX * Double(planeW)))
    let y0 = max(1, Int(faceRect.minY * Double(planeH)))
    let x1 = min(planeW - 2, Int(faceRect.maxX * Double(planeW)))
    let y1 = min(planeH - 2, Int(faceRect.maxY * Double(planeH)))
    guard x1 > x0, y1 > y0 else { return (0.5, 0.0) }

    var lumaSum = 0
    var lumaCount = 0
    let brightnessStride = 8
    var py = y0
    while py <= y1 {
      var px = x0
      while px <= x1 {
        lumaSum += Int(yPtr[py * bpr + px])
        lumaCount += 1
        px += brightnessStride
      }
      py += brightnessStride
    }
    let brightness = lumaCount > 0 ? Double(lumaSum) / Double(lumaCount) / 255.0 : 0.5

    var lapValues = [Double]()
    let lapStride = 16
    py = y0
    while py <= y1 {
      var px = x0
      while px <= x1 {
        let c = Int(yPtr[py * bpr + px])
        let lap = Double(
          -4 * c +
            Int(yPtr[py * bpr + px - 1]) +
            Int(yPtr[py * bpr + px + 1]) +
            Int(yPtr[(py - 1) * bpr + px]) +
            Int(yPtr[(py + 1) * bpr + px])
        )
        lapValues.append(lap)
        px += lapStride
      }
      py += lapStride
    }

    var sharpness = 0.0
    if lapValues.count > 1 {
      let mean = lapValues.reduce(0, +) / Double(lapValues.count)
      let variance = lapValues.map { ($0 - mean) * ($0 - mean) }.reduce(0, +) / Double(lapValues.count)
      sharpness = min(variance / CaptureConfig.sharpnessBaseline, 1.0)
    }

    return (brightness, sharpness)
  }
}

File: modules/FaceDetection/android/src/main/java/com/margelo/nitro/facedetection/FaceDetectionFrameProcessor.kt
package com.margelo.nitro.facedetection

import android.graphics.ImageFormat
import androidx.camera.core.ImageProxy
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.face.Face
import com.google.mlkit.vision.face.FaceDetection
import com.google.mlkit.vision.face.FaceDetectorOptions
import com.margelo.nitro.NitroModules
import com.margelo.nitro.camera.HybridFrameSpec
import com.margelo.nitro.camera.public.NativeFrame
import java.util.concurrent.atomic.AtomicReference
import kotlin.math.max
import kotlin.math.min
import kotlin.math.pow

class FaceDetectionFrameProcessor : HybridFaceDetectionFrameProcessorSpec() {
  private val detector = FaceDetection.getClient(
    FaceDetectorOptions.Builder()
      .setPerformanceMode(FaceDetectorOptions.PERFORMANCE_MODE_FAST)
      .setLandmarkMode(FaceDetectorOptions.LANDMARK_MODE_NONE)
      .setClassificationMode(FaceDetectorOptions.CLASSIFICATION_MODE_NONE)
      .setContourMode(FaceDetectorOptions.CONTOUR_MODE_NONE)
      .enableTracking()
      .setMinFaceSize(0.10f)
      .build(),
  )

  private val appContext = NitroModules.applicationContext?.applicationContext
    ?: throw IllegalStateException("NitroModules.applicationContext is null; FaceDetection cannot initialize.")
  private val pipeline = FaceCapturePipeline(appContext)

  private val pendingCapture = AtomicReference<CapturePayload?>(null)
  private val lastTargetPose = AtomicReference<String?>(null)

  init {
    pipeline.onCapture = { payload -> pendingCapture.set(payload) }
  }

  override fun processFrame(
    frame: HybridFrameSpec,
    args: FaceDetectionArgs,
  ): Variant_NullType_GuidanceResult_CaptureResult {
    val targetPose = args.targetPose
    val previousPose = lastTargetPose.getAndSet(targetPose)
    if (previousPose != null && previousPose != targetPose) {
      // Pose step changed (e.g. center -> left). Drop any stale capture from the
      // previous step and restart stabilization for the new target.
      pendingCapture.set(null)
      pipeline.reset()
    }

    pendingCapture.getAndSet(null)?.let { payload ->
      // Drop stale captures from a previous pose step.
      if (payload.poseId == targetPose) {
        return Variant_NullType_GuidanceResult_CaptureResult.create(
          CaptureResult(
            type = NativeResultType.CAPTURED,
            poseId = payload.poseId,
            uri = payload.uri,
            qualityScore = payload.qualityScore,
            scores = CaptureScores(
              brightness = payload.scores.brightness.toDouble(),
              sharpness = payload.scores.sharpness.toDouble(),
              centeredness = payload.scores.centeredness.toDouble(),
              poseAccuracy = payload.scores.poseAccuracy.toDouble(),
              stability = payload.scores.stability.toDouble(),
              faceSize = payload.scores.faceSize.toDouble(),
              composite = payload.scores.composite.toDouble(),
            ),
          ),
        )
      }
    }

    val imageProxy = (frame as? NativeFrame)?.image ?: return noFaceResult(targetPose)
    val frameW = imageProxy.width.toFloat()
    val frameH = imageProxy.height.toFloat()

    // ML Kit's fromMediaImage only accepts YUV_420_888 or JPEG.
    // For RGBA_8888 or any other format produced by HybridFrameOutput, fall back
    // to fromBitmap so detection always works regardless of camera output format.
    val mediaImage = imageProxy.image
    val inputImage = try {
      if (mediaImage != null && imageProxy.format == ImageFormat.YUV_420_888) {
        InputImage.fromMediaImage(mediaImage, imageProxy.imageInfo.rotationDegrees)
      } else {
        InputImage.fromBitmap(imageProxy.toBitmap(), imageProxy.imageInfo.rotationDegrees)
      }
    } catch (_: Exception) {
      return noFaceResult(targetPose)
    }
    val faces: List<Face> = try {
      com.google.android.gms.tasks.Tasks.await(
        detector.process(inputImage),
        300,
        java.util.concurrent.TimeUnit.MILLISECONDS,
      )
    } catch (_: Exception) {
      return noFaceResult(targetPose)
    }

    val face = faces.firstOrNull() ?: return noFaceResult(targetPose)

    val rect = face.boundingBox
    val rotation = imageProxy.imageInfo.rotationDegrees
    val (normW, normH) = if (rotation == 90 || rotation == 270) {
      frameH.toDouble() to frameW.toDouble()
    } else {
      frameW.toDouble() to frameH.toDouble()
    }
    val bbX = (rect.left.toDouble() / normW).coerceIn(0.0, 1.0)
    val bbY = (rect.top.toDouble() / normH).coerceIn(0.0, 1.0)
    val bbW = (rect.width().toDouble() / normW).coerceIn(0.0, 1.0)
    val bbH = (rect.height().toDouble() / normH).coerceIn(0.0, 1.0)
    val cx = (rect.exactCenterX().toDouble() / normW).coerceIn(0.0, 1.0).toFloat()
    val cy = (rect.exactCenterY().toDouble() / normH).coerceIn(0.0, 1.0).toFloat()
    val faceSizeRatio = bbH.toFloat()

    val rawYaw = face.headEulerAngleY
    val yaw = if (CaptureConfig.invertYawForFrontCamera) -rawYaw else rawYaw
    // Normalize pitch to user-facing semantics: positive = UP, negative = DOWN.
    // ML Kit's headEulerAngleX already follows this convention on Android.
    val pitch = face.headEulerAngleX
    val roll = face.headEulerAngleZ.toDouble()

    val (rawBrightness, rawSharpness) = qualityMetrics(
      imageProxy,
      frameW.toInt(),
      frameH.toInt(),
      bbX,
      bbY,
      bbW,
      bbH,
    )

    val progress = pipeline.processFrame(
      image = imageProxy,
      faceDetected = true,
      yaw = yaw,
      pitch = pitch,
      cx = cx,
      cy = cy,
      faceSizeRatio = faceSizeRatio,
      rawBrightness = rawBrightness.toFloat(),
      rawSharpness = rawSharpness.toFloat(),
      targetPose = targetPose,
    )

    return Variant_NullType_GuidanceResult_CaptureResult.create(
      GuidanceResult(
        type = NativeResultType.GUIDANCE,
        faceDetected = true,
        boundingBoxX = bbX,
        boundingBoxY = bbY,
        boundingBoxWidth = bbW,
        boundingBoxHeight = bbH,
        yaw = yaw.toDouble(),
        pitch = pitch.toDouble(),
        roll = roll,
        brightness = rawBrightness,
        sharpness = rawSharpness,
        faceSizeRatio = faceSizeRatio.toDouble(),
        faceCenterX = cx.toDouble(),
        faceCenterY = cy.toDouble(),
        stabilizationProgress = progress.toDouble(),
      ),
    )
  }

  private fun qualityMetrics(
    image: ImageProxy,
    frameW: Int,
    frameH: Int,
    bbX: Double,
    bbY: Double,
    bbW: Double,
    bbH: Double,
  ): Pair<Double, Double> {
    // Y-plane analysis requires YUV_420_888 and rotation=0/180.
    // For rotated frames (portrait phones: rotation=90/270) the bounding box is in
    // display-space but Y-plane indices are in sensor-space — axes are transposed,
    // so we'd read from the wrong region. Return neutral values instead.
    if (image.format != ImageFormat.YUV_420_888) return 0.5 to 0.5
    val rotation = image.imageInfo.rotationDegrees
    if (rotation == 90 || rotation == 270) return 0.5 to 0.5
    val yPlane = image.planes[0]
    val yBuf = yPlane.buffer
    val rowStride = yPlane.rowStride
    val pixStride = yPlane.pixelStride

    val x0 = max(1, (bbX * frameW).toInt())
    val y0 = max(1, (bbY * frameH).toInt())
    val x1 = min(frameW - 2, ((bbX + bbW) * frameW).toInt())
    val y1 = min(frameH - 2, ((bbY + bbH) * frameH).toInt())
    if (x1 <= x0 || y1 <= y0) return 0.5 to 0.0

    fun luma(x: Int, y: Int) = yBuf.get(y * rowStride + x * pixStride).toInt() and 0xFF

    var sum = 0L
    var n = 0
    val brightnessStride = 8
    var py = y0
    while (py <= y1) {
      var px = x0
      while (px <= x1) {
        sum += luma(px, py)
        n++
        px += brightnessStride
      }
      py += brightnessStride
    }
    val brightness = if (n > 0) sum.toDouble() / n / 255.0 else 0.5

    val laps = ArrayList<Double>()
    val lapStride = 16
    py = y0
    while (py <= y1) {
      var px = x0
      while (px <= x1) {
        laps.add((-4.0 * luma(px, py) + luma(px - 1, py) + luma(px + 1, py) + luma(px, py - 1) + luma(px, py + 1)))
        px += lapStride
      }
      py += lapStride
    }
    val sharpness = if (laps.size > 1) {
      val mean = laps.average()
      (laps.map { (it - mean).pow(2) }.average() / CaptureConfig.sharpnessBaseline).coerceIn(0.0, 1.0)
    } else {
      0.0
    }

    return brightness to sharpness
  }

  private fun noFaceResult(targetPose: String = "center"): Variant_NullType_GuidanceResult_CaptureResult {
    pipeline.processFrame(null, false, 0f, 0f, 0.5f, 0.5f, 0f, 0.5f, 0f, targetPose)
    return Variant_NullType_GuidanceResult_CaptureResult.create(
      GuidanceResult(
        type = NativeResultType.GUIDANCE,
        faceDetected = false,
        boundingBoxX = null,
        boundingBoxY = null,
        boundingBoxWidth = null,
        boundingBoxHeight = null,
        yaw = null,
        pitch = null,
        roll = null,
        brightness = null,
        sharpness = null,
        faceSizeRatio = null,
        faceCenterX = null,
        faceCenterY = null,
        stabilizationProgress = 0.0,
      ),
    )
  }
}
